File: day01/day01.py
## ################################## ##
#### ### ### Advent of Code ### ### ####
# www.github.com/TeeJizzm/AdventOfCode #
## ################################## ##

## Day 01

############################
# Imports

import os

import tools.texttolists as tl

############################
# Variables



############################
# Functions

def day01(text):
    print("Day 01 - Historian Hysteria")
    
    part1, part2 = 0, 0
    
    pairs = tl.toTupleList(text,"\n","   ")
    lists = tl.to2dArray(text,"\n","   ")

    list_l = sorted([i[0] for i in pairs])
    list_r = sorted([i[1] for i in pairs])

    for i in range(len(pairs)):
        diff = abs(int(list_l[i]) - int(list_r[i]))
        part1 += diff

    for num in set(list_l):
        cr = list_r.count(num)
        cl = list_l.count(num)

        part2 += cr * cl * int(num)

    return part1, part2

############################
# Run main

if __name__ == "__main__":
    day = "day01/inc" 
    
    # Change file
    #######
    #file = "ex.txt"
    file = "in.txt"
    #######
    
    # Get absolute filepath of file
    filepath = os.path.join(os.getcwd(), day, file)
    
    # Open file, clean up memory after
    with open(filepath, "r") as file:
        
        text = file.read() # Read data

    part1, part2 = day01(text)

    print("Part 1: ", part1)
    print("Part 2: ", part2)
    
########### EOF ############

File: day01/day01.go
/* ################################## ##
#### ### ### Advent of Code ### ### ####
# www.github.com/TeeJizzm/AdventOfCode #
## ################################## */

// Day 01

// -----------------------------------

package main

// Imports
import (
	// Input/Output
	"fmt"
	"os"

	// String manipulation
	"strings"
	// Other required import
	"slices"
)

// -----------------------------------
// Functions //

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func countOcc(slice []int, target int) int {
	count := 0
	for _, value := range slice {
		if value == target {
			count++
		}
	}
	return count
}

func part_one(col_a []int, col_b []int) int {
	var part1 int = 0

	//fmt.Println(col_a)
	//fmt.Println(col_b)

	for i, a := range col_a {
		diff := a - col_b[i]

		part1 += max(diff, -diff)
	}

	return part1
}

func part_two(col_a []int, col_b []int) int {
	var part2 int = 0

	set_a := make([]int, len(col_a))
	copy(set_a, col_a)

	slices.Compact(set_a)

	for _, num := range set_a {
		cnt_a := countOcc(col_a, num)
		cnt_b := countOcc(col_b, num)

		val := cnt_a * cnt_b * num

		part2 += val
	}

	return part2
}

func main() {
	fmt.Println("Day 01 - Historian Hysteria")

	// ------------ File -------------

	/* <- File switch
	file, err := os.ReadFile("day01/inc/ex.txt")
	/*/ //
	file, err := os.ReadFile("day01/inc/in.txt")
	// --- End --- */

	check(err)

	// Added fix for Windows newline
	text := strings.ReplaceAll(string(file), "\r\n", "\n")
	//fmt.Println(text)

	// ----------- Setup -------------

	lines := strings.Split(text, "\n")

	col_a := make([]int, len(lines))
	col_b := make([]int, len(lines))

	for i, line := range lines {
		var a, b int
		_, err = fmt.Sscan(line, &a, &b)
		check(err)

		col_a[i] = a
		col_b[i] = b
	}

	slices.Sort(col_a)
	slices.Sort(col_b)

	part1 := part_one(col_a, col_b)
	part2 := part_two(col_a, col_b)

	// ----------- Output ------------

	fmt.Println("Part 1:", part1)
	fmt.Println("Part 2:", part2)
}

// -----------------------------------
